src/lib/SaferNames/Name.hs

```diff
@@ -316,168 +234,339 @@ data ListE (e::E) (n::S) = ListE { fromListE :: [e n] }
+
+applyNaryAbs :: Typeable s => Typeable val
+             => ScopeReader m
+             => BindsNameList b s
+             => SubstE (SubstVal s val) e
+             => SubstB (SubstVal s val) b
+             => Abs b e n -> [val n] -> m n (e n)
+applyNaryAbs (Abs bs body) xs = applySubst (idSubst <>> (bs @@> map SubstVal xs)) body
+
+-- === versions of monad constraints with scope params ===
+
+type MonadKind  =           * -> *
+type MonadKind1 =      S -> * -> *
+type MonadKind2 = S -> S -> * -> *
+
+type Monad1 (m :: MonadKind1) = forall (n::S)        . Monad (m n  )
+type Monad2 (m :: MonadKind2) = forall (n::S) (l::S) . Monad (m n l)
+
+type MonadErr1 (m :: MonadKind1) = forall (n::S)        . MonadErr (m n  )
+type MonadErr2 (m :: MonadKind2) = forall (n::S) (l::S) . MonadErr (m n l)
+
+type MonadFail1 (m :: MonadKind1) = forall (n::S)        . MonadFail (m n  )
+type MonadFail2 (m :: MonadKind2) = forall (n::S) (l::S) . MonadFail (m n l)
+
+-- === scope reader monad ===
+
+class Monad1 m => ScopeReader (m::MonadKind1) where
+  askScope :: m n (Scope n)
+  extendScope :: ScopeFrag n l -> m l a -> m n a
+
+type ScopeReader2 (m::MonadKind2) = forall (n::S). ScopeReader (m n)
+
+newtype ScopeReaderT (m:: * -> *) (n::S) (a:: *) =
+  ScopeReaderT { runScopeReaderT' :: ReaderT (Scope n) m a}
+  deriving (Functor, Applicative, Monad)
+
+type ScopeReaderM = ScopeReaderT Identity
 
-extendScope :: HasNamesB b => Scope n -> b n l -> Scope l
-extendScope s b = s <>> boundScope b
+runScopeReaderT :: Monad m => Scope n -> ScopeReaderT m n a -> m a
+runScopeReaderT scope m = runReaderT (runScopeReaderT' m) scope
 
-freshenAbs :: (HasNamesB b, HasNamesE e) => Scope n -> Abs b e n -> FreshAbs b e n
-freshenAbs s (Abs b body) =
-  case freshenBinder s b of
-    FreshBinder ext b' renamer -> do
-      let s' = extendScope s b'
-      let t = extendInjectNameTraversal ext renamer idNameTraversal
-      let body' = runIdentity $ traverseNamesE s' t body
-      FreshAbs ext b' body'
+runScopeReaderM :: Scope n -> ScopeReaderM n a -> a
+runScopeReaderM scope m = runIdentity $ runScopeReaderT scope m
 
-freshenBinder :: HasNamesB b => Scope n -> b n l -> FreshBinder b n (n:=>:l)
-freshenBinder s b = runIdentity $ traverseNamesB s idNameTraversal b
+lookupScopeM :: Typeable s => ScopeReader m => Name s n -> m n (s n)
+lookupScopeM v = flip lookupScope v <$> askScope
 
-extendRecEnv :: HasNamesE e => FreshExt n l -> RecEnv e n -> RecEnvFrag e n l -> RecEnv e l
-extendRecEnv ext (RecEnv env) (RecEnvFrag frag) = let
-  env' = injectEnvNamesL ext env
-  in RecEnv $ env' <>> frag
+withDistinct :: ScopeReader m => (Distinct n => m n a) -> m n a
+withDistinct cont = askScope >>= \case Scope _ -> cont
 
-injectEnvNamesL :: HasNamesE e => FreshExt n l -> Env i (e n) -> Env i (e l)
-injectEnvNamesL ext env = fromEnvE $ injectNamesL ext $ EnvE env
+-- === subst monad ===
+
+class ScopeReader2 m => SubstReader (v::E->E) (m::MonadKind2) | m -> v where
```

Comment:
At this point I am completely lost 😕  I need more docs/explanation

Reply:
Sorry :/ . Let's chat and figure out which are the tricky parts that require extra docs